framework/src/systems/NonlinearSystemBase.C

```diff
@@ -1084,6 +1085,65 @@ NonlinearSystemBase::enforceNodalConstraintsJacobian()
   }
 }
 
+void
+NonlinearSystemBase::reinitNodeFace(const Node & secondary_node,
+                                    const BoundaryID secondary_boundary,
+                                    const PenetrationInfo & info,
+                                    const bool displaced)
+{
+  auto & subproblem = displaced ? static_cast<SubProblem &>(*_fe_problem.getDisplacedProblem())
+                                : static_cast<SubProblem &>(_fe_problem);
```

Comment:
The `displaced == false` version of subproblem is never used. @lindsayad do we have design docs on when to use the undisplaced problem even if we have a displaced problem? I think in theory we are duplicating a lot of the bookkeeping in the two problems, but I'm guessing we only care about the undisplaced problem being up-to-date? Is there anything in the code that prevents devs from accidentally accessing state that is not being maintained on the displaced problem?

Reply:
It wasn't until this PR that we started doing dedicated reinits on `_fe_problem`. Previously we just used the `subproblem` corresponding to the value of `displaced`. I presume that @recuero made this change because he wanted to use some kind of evaluations (materials maybe?) from the undisplaced problem even in displaced node-face constraints

Reply:
My concern now, which I wasn't really that motivated to voice until your comment here, is the *reverse*. If we are going to allow use of undisplaced calculations in displaced node-face constraints, why not allow the reverse? E.g. perhaps we should also always reinit the displaced problem even if `displaced == false`

Reply:
I should be more clear. We used to only do `subproblem.reinitNeighborPhys`. (We have always done _fe_problem.reinitNodeFace`). The change that this PR introduced is that we now always do `_fe_problem.reinitNeighborPhys` and then if `displaced == true` we also do `subproblem.reinitNeighborPhys` (corresponding to the displaced problem)

Reply:
>  I presume that @recuero made this change because he wanted to use some kind of evaluations (materials maybe?) from the undisplaced problem even in displaced node-face constraints

That's exactly right.

The case for doing displaced mesh evaluations for an undisplaced object is less straightforward to me. I.e. maybe less likely to happen? The motivation here was to avoid failed mappings.

Reply:
Yes, I agree it's probably not very likely